### ISSUE TITLE
[Feature request] Unsynced units/features

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/spring#644]
# User story, use case etc

Trees and rocks are visible in fog of war (via the feature visibility modrule), but this lets players see when they get reclaimed or crushed. I want them to appear intact until the player actually sees them. This is what ghosts generally do. But ghosts are bad for this case: they look quite different, they aren't clickable or have tooltip, and it's harder to make them work with things that call `Spring.GetAllFeatures()` (or visible features, or in rectangle, doesn't matter) and work on that set, for example reclaim value summary, and it wouldn't be passed to callins like DrawFeatureBla, for example for the swaying tree shader. I want something that behaves like a feature in the ways mentioned above, but is still unsynced, i.e. can be created and manipulated at will and doesn't affect other players.

This could be extended to units, where the benefit is much greater because units have order queues (which would be translated into something else, since the unit wouldn't act on its queue). Here are some ways one could use unsynced units:

1) initial queue and startpoint. Currently the widget doesn't implement all commands (can't jump or right-click to move) and assumes only one commander (so if you get a double commander then one gets the commands and one is left idle). If the commanders were unsynced units you could select them and give them commands using native engine UI instead of having a broken-ass emulation.

2) shared factory rally queues. If you want a second factory to have the same rally queue as the first, you need to replicate the queue. You could probably have some special "replicate queue" command but that still leaves one factory as the primary (what if it dies? Meanwhile an unsynced unit can't die and can be distinguished from one of N factories more easily). Similarly it could represent a ferry queue. Existing ZK ferry queue is quite limited (only handles move commands, requires special UI to give orders, looks different compared to regular queues), being able to use a native order queue would be a great UX improvement. Mockup screenshot (low visibility but the second factory is guarding a fake unit with a bunch of fight commands queued):
![image](https://user-images.githubusercontent.com/2573076/219027378-a106d721-7c6d-4e6a-b595-a7c4353fe606.png)

3) AI strategic command. You would have a unit representing an allied AI player and commands to the unit would be reinterpreted as requests for the AI to do something in a specific place. This is similar to the Red Alert 3 campaign UI where you can direct your AI ally, but again the benefit here is the UX of reusing native orders UI. Here's a mockup where the player directs an AI sitting in the middle to go through the bottom instead via an unsynced control unit:
![image](https://user-images.githubusercontent.com/2573076/219027822-30cecb2c-1536-4348-8d54-686b44e6768f.png)

4) Warcraft style shops. Select the shop like a unit, rightclick to select buying unit, have items listed as commands on the command panel. This could also be done by having the shop be a real unit and sharing control, but that's an alternative with various pros and cons.
![image](https://user-images.githubusercontent.com/2573076/219031078-e33f3bab-2af9-4c26-8b74-697250cc8778.png)

5) limited control of ally units, in particular ZK strider hub would have an unsynced unit underneath that could allow placing nanoframes. Again, doable by allowing selection of the real unit but it's a different tradeoff.
![image](https://user-images.githubusercontent.com/2573076/219031984-fe2de025-df0a-4804-b53e-86a4c7e0dc25.png)

6) get native tooltip and right-clickability on things like existing ghost site widgets or lingering ground AoE effects.

## Technicalities

This would be a very large task and a lot can go wrong (since everything currently assumes units and features are synced). Additionally there are many questions where the answer isn't obvious yet (if I select all units of a type by Ctrl+Z, does it select the unsynced ones as well? what does a widget do if none of the unitDefs provided by the game suffice for its purpose? etc). Some questions have seemingly obvious answers (what about featureID number pool? -> since the IDs should never be sent on network it can be drawn from a separate ID pool over 2^16) but reality will likely verify this. So perhaps if this gets taken up it would be ideal to do in parts so proof of concept could be made an issues discovered. Here's a suggestion for a bunch of specific implementation details:

 * add `Spring.CreateUnsyncedFeature(featureDef, x, y, z, ...)` with the same interface as `CreateFeature` to UnsyncedCtrl. The name doesn't matter much, maybe it could have just `CreateFeature` since some unsynced functions already differ under the same name (for example GiveOrderToUnit), or maybe it could be something like CreateVirtualFeature, doesn't matter.
 * add `Spring.SetUnsyncedFeatureHealth` and other control interfaces that mimic the real ones to UnsyncedCtrl. These can of course only control unsynced units.
 * make unsynced `Spring.GetFeatureHealth` return whatever health is set (note, no "unsynced" in the name, this is the existing function that works with synced features as well. But only returns anything for unsynced units in unsynced contexts.). Ditto for the other getters. In particular, GetAllFeatures (and similar that get a list) returns unsynced features too.
 * add some sort of `Spring.IsFeatureSynced(featureID) -> nil (not valid at all), true (is synced), or false (unsynced). Cannot return false if ran from a synced context
 * drawn normally and passed to `wupget:DrawFeatureBla`.
 * intercepted by mouse raytraces, so has a tooltip, is passed to `wupget:DefaultCommand` and right-clicking produces an order request. If a synced unit gets a unit-targeted order targeted onto an unsynced unit, this gets passed to UnitCommand normally (or whichever was the callin before the command gets sent over the network) at which point the code should probably translate it to a real command; then the order gets dropped.
 * is not subject to updates (so does not follow physics, does not do things like smoking if it's an unsynced geovent etc)
 * is not counted in synced area-based code like AoE damage, does not respond to synced raytraces, etc

If that turns out usable then maybe it can get extended to units in a similar way. The main thing units do that features don't is that they have commandAI - they should produce a UnitCommand event but (regardless of whether the target is synced or another unsynced units) it goes directly to the unit's actual queue (as opposed to synced units that take a network roundtrip when given orders from unsynced; similar to how synced GiveOrder works) and the unit also doesn't act on the queue. Units also have a script, I'm not sure how to approach this but that is something to think about later.

**Comment by Krogoth100** [comment]
This is an interesting stuff, but I am myself gonna experiment with it in a few upcoming months. My approach is a bit different, I generate visual objects entirely on GPU. Such objects are put on a VBO and are operated by a compute shader. List of visible objects is also created on and by GPU itself, and then used with an indirect draw call (unavailable yet, afaik, but it's not a blocker).
I have no idea how well this is, but it kinda makes sense to do this stuff on GPU. We can benchmark it billion times and on different machines it could different results; the idea is not about performance, its about where this execution code belongs: CPU or GPU.
In my current understanding, we should do such stuff purely GPU.
Consider this option.

**Comment by sprunk** [comment]
This request is not about rendering in general, and these are not just visual objects. In the context of this request I entirely don't care if the engine renders things on GPU or the CPU.

The request is about not having to make imperfect gameside reimplementations of things the engine already does (rendering being one of those, but also mouse-over handling including selection and tooltip, order queues, and whatnot).

The engine exposes some tools to do most of these things in isolation. For rendering that's your VBO, or maybe `gl.Unit`; for mouse-over tooltip, there's `Spring.TraceRay`; for orders you can populate whatever widget displays the command panel with fake values and intercept clicks; etc. The point is that sometimes you want them all together, and you want the behaviour of your emulation to exactly match what the engine does up to the little details, and you want this without going through great lengths to both write all these systems and then maintain behaviour identical with the engine in case something changes over time.

**Comment by Krogoth100** [comment]
I understand.
I just offer some thoughts about where I have some experience.
"Ghosts" _and_ "warcraft shop". Both seem purely visual, with exception of some proximity synced check.
I have done custom factory units. Including stuff like mobile factories like SupCom's Fatboy and such, which are factories but not factories. These are getting well-coded in Lua, because they are not spam units, Lua performance cost doesn't matter to them.
About replicated factory queues. If factory dies, can't you just relegate its queue to the next alive one?
About morphing units: I think engine should explicitly allow morphing unit to receive commands of a result unit type. We have same feature for unit's that are not built yet, why should it be different for morphing. This seems like a proper solution for me.

I am not trying to say this unsynced object idea is bad or anything like that, it just looks very generalized. In my opinion, you can eliminate few cases and get a more clear picture.

**Comment by sprunk** [comment]
> "Ghosts" and "warcraft shop". Both seem purely visual, with exception of some proximity synced check.

Current ghosts are only visual and that's one of my issues with them. I want them returned by getters, I want them to have a tooltip, I want to right-click them etc. and I don't want to reimplement all that gameside.

Warcraft shop is not about the visuals, it's about offering selection-based UI to multiple players at once.

> I have done custom factory units.

I want a regular, plain factory to have two queues (and I don't want to reimplement queue handling from scratch).

> About replicated factory queues. If a factory dies, can't you just relegate its queue to the next alive one?

No, because that arbitrarily shuffles the player's input point and can lose it completely if all factories die. I want the player to be able to always select and manipulate a permanent thing representing the rally queue itself.

> About morphing units: I think engine should explicitly allow morphing unit to receive commands of a result unit type. We have same feature for unit that are not built yet, why should it be different for morphing?

A unit being built already has the correct type (unitDef) and a morphing unit does not. The engine cannot explicitly allow morphing units to do anything because morphing is not a native engine concept, it's entirely gameside. The engine can only expose tools for the game to implement what it wants. Regardless, part of it is already doable with current API (you can add commandDescs to the unit and clear them if morphing fails) and part of it is not (what if the morph splits the unit into two?) but again the point is that I don't even want to write the doable part, the engine already handles the default command sets for unit types natively.

**Comment by Krogoth100** [comment]
Yeah, I want that for ghosts too. I mean, these "visual" objects are totally gettable, its just the work is done inside a shader. Shader can trace an object and spawn a tooltip if needed. It's just compute shaders doing such work are quite quick, that's why I am considering it. I mean, if you go this route, you don't need engine changes to do it. I have seen examples of GUI implemented almost entirely via shaders, so such works exist. (_Once again, it's not about not touching engine and finding out workaround. It's about extended usage of modern graphic features, that enable to do some new stuff._)
Same goes for shop, shader picks up a shop and spawns a menu of items. There are some concerns about proper interactions such as a mouse click without the delay, but I think those are solvable.
So basically you have a bunch of shaders that do the work steadily and a small Lua interface to control the state of such objects.
Normal, physical objects already exist for shaders and are put on a UBO and SSBO buffers. These objects I am talking about exist _only_ on SSBO buffers, do not require to be present on a CPU side.
I mean, I am gonna do all that^ in future, but I have nothing to show for now.

As for control objects, such that factory queues and morphing units, I agree it could be useful of course, why not. Warcraft 3 (and Starcraft 2 I presume) custom maps extensively used so called "dummy" units. Its a normal practise, it seems.

But these 2 cases, I perceive them separately.

**Comment by sprunk** [comment]
> these "visual" objects are totally gettable, its just the work is done inside a shader. (...) Normal, physical objects already exist for shaders and are put on a UBO and SSBO buffers. These objects I am talking about exist only on SSBO buffers, do not require to be present on a CPU side.

I don't just want any arbitrary getting, I want them specifically to work with existing interfaces (in particular, `Spring.GetFeatureX`, and `Spring.GetXFeatures` and `wupget:DrawFeatureX`). Making these work on the GPU in the background for unsynced Lua states (or unsynced engine state in general) sounds like a fine feature request, but I only care about the interface, not the implementation.

> Same goes for shop, shader picks up a shop and spawns a menu of items. There are some concerns about proper interactions such as a mouse click without the delay, but I think those are solvable. (...) I mean, if you go this route, you don't need engine changes to do it.

I don't want to bother myself with those concerns. This is explicitly a request for an engine feature because I don't want to go the "reimplement everything gameside" route (regardless of how I would implement it if I were in fact doing it gameside).

> I am gonna do all that^ in future, but I have nothing to show for now.

Cool, thanks! If you show something robust maybe I would consider it good enough to reuse and some aspects of this request would become moot.

> Warcraft 3 (and Starcraft 2 I presume) custom maps extensively used so called "dummy" units. Its a normal practise, it seems.

I don't have that much experience with the WC3 engine but as far as I can tell those dummies are synced. Synced dummies are already possible in our engine. But being synced makes them unsuitable for many uses:
 * you need to exempt them from most game mechanics (for example "game over if you have 0 units", this needs to skip dummies) and allied things (for example "how many units does my ally have" UI). This is doable, but is needless work. Unsynced dummies would be exempt from most mechanics by default.
 * anybody working with your game (AI developers, players making widgets) also has to remember to do the above which turns a technical problem partially social.
 * bad scaling. If you have a 16v16 you would be spawning 15 extra dummies for things like allied control. If you spawn N dummies for extra factory rally queues, what if the user wants N+1? Solutions can introduce further problems (for example requesting to spawn a new dummy would have to take into account the network roundtrip delay). It may be solvable to a reasonable extent but I don't want to be solving those problems.

**Comment by Krogoth100** [comment]
Synced dummies are out of question, sure.
